python/tournament-scheduling/src/tournament_scheduling/domain.py

```diff
@@ -0,0 +1,97 @@
+from timefold.solver import SolverStatus
+from timefold.solver.domain import (planning_entity, planning_solution, PlanningId, PlanningVariable,
+                                    PlanningEntityCollectionProperty,
+                                    ProblemFactCollectionProperty, ValueRangeProvider,
+                                    PlanningScore, PlanningPin)
+from typing import Dict, List, Any, Annotated
+
+from .json_serialization import *
+
+
+class Day(JsonDomainBase):
+    date_index: int
+
+    def __eq__(self, other):
+        if self is other:
+            return True
+        if not isinstance(other, Day):
+            return False
+        return self.date_index == other.date_index
+
+    def __hash__(self):
+        return 31 * self.date_index
+
+
+class Team(JsonDomainBase):
+    id: Annotated[int, PlanningId]
+    name: Annotated[str, Field(default=None)]
+
+
+    def __eq__(self, other):
+        if self is other:
+            return True
+        if not isinstance(other, Team):
+            return False
+        return self.id == other.id
+
+    def __hash__(self):
+        return 31 * hash(self.id)
+
+    def __str__(self):
+        return self.name if self.name is not None else super().__str__()
+
+    def __repr__(self):
+        return f'Team({self.id}, {self.name})'
+
+
+class UnavailabilityPenalty(JsonDomainBase):
+    team: Annotated[Team | None,
+                   IdSerializer,
+                   TeamDeserializer,
+                   Field(default=None)]
+    day: Annotated[Day | None,
+                  IdSerializer,
+                  DayDeserializer,
+                  Field(default=None)]
+
+
+@planning_entity
+class TeamAssignment(JsonDomainBase):
+    id: Annotated[int, PlanningId]
+    day: Annotated[Day | None,
+                  IdSerializer,
+                  DayDeserializer,
+                  Field(default=None)]
+    index_in_day: int
+    pinned: Annotated[bool, PlanningPin]
```

Comment:
I would make this default to `False` (in Java, all `boolean` values are initialized to `false` even if they don't appear in a constructor and don't have a field initializer.

```suggestion
    pinned: Annotated[bool, PlanningPin, Field(default=False)]
```